src/main/java/co/cask/hydrator/python/transform/Py4jPythonExecutor.java

```diff
@@ -0,0 +1,230 @@
+              .until(() -> (initializedPythonProcess()));
+    } catch (ConditionTimeoutException e) {
+      String message =  String.format("Connection to python process failed in %d seconds.\n%s",
+              PY4J_INIT_TIMEOUT, getProcessOutput());
+      throw new RuntimeException(message);
+    }
+  }
+
+  @Override
+  public void transform(StructuredRecord input, final Emitter<StructuredRecord> emitter,
+                        Emitter<Map> pythonEmitter, ScriptContext scriptContext) {
+    try {
+      Object record = PythonObjectsEncoder.encode(input, input.getSchema());
+      py4jTransport.transform(record, pythonEmitter, scriptContext);
+    } catch (Exception e) {
+      if (e instanceof Py4JException) {
+        logger.error(e.getMessage());
+      }
+      throw new IllegalArgumentException("Could not transform input.\n" + e.getMessage());
+    }
+  }
+
+  @Override
+  public void destroy() {
+    try {
+      if (py4jTransport != null) {
+        py4jTransport.finish();
+      }
+    } catch (Exception e) {
+      logger.warn("Cannot close python process.\n", e);
+    }
+
+    if (gatewayServer != null) {
+      gatewayServer.shutdown();
+    }
+
+    logger.debug("Waiting for python process to finish");
+
+    if (process == null) {
+      logger.debug("Process was not created. Not waiting for it to finish.");
+      return;
+    }
+
+    try {
+      process.waitFor(PYTHON_PROCESS_STOP_TIMEOUT, SECONDS);
+      process.destroy(); // tell the process to stop
+      process.waitFor(PYTHON_PROCESS_KILL_TIMEOUT, SECONDS); // give it a chance to stop
+      process.destroyForcibly();
+    } catch (InterruptedException e) {
+      int overallTimeout = PYTHON_PROCESS_STOP_TIMEOUT + PYTHON_PROCESS_KILL_TIMEOUT;
+      logger.warn("Python process was not able to stop in {} seconds.", overallTimeout);
+    }
+
+    logger.info("Python process output:\n{}", getProcessOutput());
+
+    try {
+      FileUtils.deleteDirectory(transformTempDir);
+    } catch (IOException e) {
+      logger.warn("Cannot delete tmp directory {}", transformTempDir, e);
+    }
+  }
+
+  private String getProcessOutput() {
+    Path pythonProcessOutputFilePath = Paths.get(pythonProcessOutputFile.getPath());
+
+    try {
+      String processOutput = new String(Files.readAllBytes(pythonProcessOutputFilePath), StandardCharsets.UTF_8);
+      return processOutput;
+    } catch (IOException e) {
+      logger.warn("Cannot read output of python process.\n", e);
+    }
+    return null;
+  }
+
+  private boolean portFileCreated(File portFile) {
+    // usually happens if code has a syntax error
+    if (!process.isAlive()) {
+      throw new RuntimeException(String.format("Python process died too early.\n%s", getProcessOutput()));
+    }
+
+    return portFile.exists();
+  }
+
+  private boolean initializedPythonProcess() {
+    try {
+      py4jTransport.initialize(gatewayServer.getListeningPort());
+      return true;
+    } catch (py4j.Py4JException e) {
+    }
```

Comment:
place second return here. Avoid empty catch blocks. Also are you sure you can always hide this exception? It looks too wide.

Reply:
There is no more specific exception unfortunately which would indicate connection failure.

This method is polled until true or until timeout, so in general if something bad happens it won't get blindly ignored (timeout will fail the job still) 

> py4j.Py4JException: Error while obtaining a new communication channel
	at py4j.CallbackClient.getConnectionLock(CallbackClient.java:257) ~[py4j-0.10.8.1.jar:na]
	at py4j.CallbackClient.sendCommand(CallbackClient.java:377) ~[py4j-0.10.8.1.jar:na]
	at py4j.CallbackClient.sendCommand(CallbackClient.java:356) ~[py4j-0.10.8.1.jar:na]
	at py4j.reflection.PythonProxyHandler.invoke(PythonProxyHandler.java:106) ~[py4j-0.10.8.1.jar:na]
	at com.sun.proxy.$Proxy135.initialize(Unknown Source) ~[na:na]
	at co.cask.hydrator.python.transform.Py4jPythonExecutor.initialize(Py4jPythonExecutor.java:152) ~[1549392953470-0/:na]
	at co.cask.hydrator.python.transform.PythonEvaluator.initialize(PythonEvaluator.java:148) [1549392953470-0/:na]
	at co.cask.cdap.etl.common.plugin.WrappedTransform.lambda$initialize$3(WrappedTransform.java:72) [cdap-etl-core-6.0.0-SNAPSHOT.jar:na]
	at co.cask.cdap.etl.common.plugin.Caller$1.call(Caller.java:30) ~[cdap-etl-core-6.0.0-SNAPSHOT.jar:na]
	at co.cask.cdap.etl.common.plugin.StageLoggingCaller.call(StageLoggingCaller.java:40) ~[cdap-etl-core-6.0.0-SNAPSHOT.jar:na]
	at co.cask.cdap.etl.common.plugin.WrappedTransform.initialize(WrappedTransform.java:71) [cdap-etl-core-6.0.0-SNAPSHOT.jar:na]
	at co.cask.cdap.etl.common.plugin.WrappedTransform.initialize(WrappedTransform.java:34) [cdap-etl-core-6.0.0-SNAPSHOT.jar:na]
	at co.cask.cdap.etl.batch.mapreduce.MapReduceTransformExecutorFactory.getInitializedTransformation(MapReduceTransformExecutorFactory.java:319) ~[cdap-etl-batch-6.0.0-SNAPSHOT.jar:na]
	at co.cask.cdap.etl.batch.mapreduce.MapReduceTransformExecutorFactory.getTransformation(MapReduceTransformExecutorFactory.java:189) ~[cdap-etl-batch-6.0.0-SNAPSHOT.jar:na]
	at co.cask.cdap.etl.batch.mapreduce.MapReduceTransformExecutorFactory.getPipeStage(MapReduceTransformExecutorFactory.java:303) ~[cdap-etl-batch-6.0.0-SNAPSHOT.jar:na]
	at co.cask.cdap.etl.batch.mapreduce.MapReduceTransformExecutorFactory.create(MapReduceTransformExecutorFactory.java:220) ~[cdap-etl-batch-6.0.0-SNAPSHOT.jar:na]
	at co.cask.cdap.etl.batch.mapreduce.TransformRunner.<init>(TransformRunner.java:108) ~[cdap-etl-batch-6.0.0-SNAPSHOT.jar:na]
	at co.cask.cdap.etl.batch.mapreduce.ETLMapReduce$ETLMapper.initialize(ETLMapReduce.java:459) ~[cdap-etl-batch-6.0.0-SNAPSHOT.jar:na]
	at co.cask.cdap.etl.batch.mapreduce.ETLMapReduce$ETLMapper.initialize(ETLMapReduce.java:445) ~[cdap-etl-batch-6.0.0-SNAPSHOT.jar:na]
	at co.cask.cdap.internal.app.runtime.batch.MapperWrapper.run(MapperWrapper.java:122) ~[na:na]
	at org.apache.hadoop.mapred.MapTask.runNewMapper(MapTask.java:787) ~[org.apache.hadoop.hadoop-mapreduce-client-core-2.8.0.jar:na]
	at org.apache.hadoop.mapred.MapTask.run(MapTask.java:341) ~[org.apache.hadoop.hadoop-mapreduce-client-core-2.8.0.jar:na]
	at org.apache.hadoop.mapred.LocalJobRunner$Job$MapTaskRunnable.run(LocalJobRunner.java:270) ~[org.apache.hadoop.hadoop-mapreduce-client-common-2.8.0.jar:na]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[na:1.8.0_191]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[na:1.8.0_191]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[na:1.8.0_191]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[na:1.8.0_191]
	at java.lang.Thread.run(Thread.java:748) ~[na:1.8.0_191]
Caused by: java.net.ConnectException: Connection refused (Connection refused)
	at java.net.PlainSocketImpl.socketConnect(Native Method) ~[na:1.8.0_191]
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350) ~[na:1.8.0_191]
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206) ~[na:1.8.0_191]
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188) ~[na:1.8.0_191]
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392) ~[na:1.8.0_191]
	at java.net.Socket.connect(Socket.java:589) ~[na:1.8.0_191]
	at java.net.Socket.connect(Socket.java:538) ~[na:1.8.0_191]
	at java.net.Socket.<init>(Socket.java:434) ~[na:1.8.0_191]
	at java.net.Socket.<init>(Socket.java:244) ~[na:1.8.0_191]
	at javax.net.DefaultSocketFactory.createSocket(SocketFactory.java:277) ~[na:1.8.0_191]
	at py4j.CallbackConnection.start(CallbackConnection.java:226) ~[py4j-0.10.8.1.jar:na]
	at py4j.CallbackClient.getConnection(CallbackClient.java:238) ~[py4j-0.10.8.1.jar:na]
	at py4j.CallbackClient.getConnectionLock(CallbackClient.java:250) ~[py4j-0.10.8.1.jar:na]
	... 27 common frames omitted